File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <iostream>
#include <stack>
#include <queue>
#include <vector>
#include <algorithm>
#include <memory.h>

using namespace std;

vector<vector<int>> vertex(1001);
stack<int> dfs_stk;
queue<int> bfs_que;

bool visited[1001];
int main() {
	int n, m, v,a,b,top,next;
	cin >> n >> m >> v;
	for (int i = 0; i < m; i++) {
		cin >> a >> b;
		vertex[a].push_back(b);
		vertex[b].push_back(a);
	}
	for (int i = 0; i <= n; i++) sort(vertex[i].begin(), vertex[i].end(),greater<>());
	dfs_stk.push(v);
	bfs_que.push(v);
	while (!dfs_stk.empty()) {
		top = dfs_stk.top();
		if (visited[top]) {
			dfs_stk.pop();
			continue;
		}
		dfs_stk.pop();
		cout << top << ' ';
		visited[top] = true;
		for (int i = 0; i < vertex[top].size(); i++) {
			next = vertex[top][i];
			if (visited[next])continue;
			dfs_stk.push(next);
		}
	}
	memset(visited, 0, sizeof(visited));
	for (int i = 0; i < n; i++) sort(vertex[i].begin(), vertex[i].end());
	cout << '\n';
	while (!bfs_que.empty()) {
		top = bfs_que.front();
		bfs_que.pop();
		cout << top << ' ';
		visited[top] = true;
		for (int i = 0; i < vertex[top].size(); i++) {
			next = vertex[top][i];
			if (visited[next])continue;
			bfs_que.push(next);
			visited[next] = true;
		}
	}
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
	static List<List<Integer>> graph = new ArrayList<>();
	static int n,m,v;
	static Stack<Integer> stk = new Stack<>();
	static Queue<Integer> que = new LinkedList<>();
	static boolean[] dfs_visited = new boolean[1001];
	static boolean[] bfs_visited = new boolean[1001];
	
	
	public static void main(String args[]) throws Exception
	{	
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		n = Integer.parseInt(st.nextToken()); m = Integer.parseInt(st.nextToken());
		v = Integer.parseInt(st.nextToken());
		
		for(int i=0; i<=n; i++) {
			graph.add(new ArrayList<>());
		}
		
		int temp1,temp2;
		
		for(int i=0; i<m; i++) {
			st = new StringTokenizer(br.readLine());
			temp1 = Integer.parseInt(st.nextToken()); temp2 = Integer.parseInt(st.nextToken());
			graph.get(temp1).add(temp2); graph.get(temp2).add(temp1);
		}
		for(int i=0; i<=n; i++) {
			Collections.sort(graph.get(i), Collections.reverseOrder());
		}
		
		// dfs
		stk.push(v);
		while(!stk.empty()) {
			int top = stk.pop();
			if(!dfs_visited[top]) {
				dfs_visited[top] = true;
				System.out.print(top+" ");
			}
			for(Integer t:graph.get(top)) {
				if(!dfs_visited[t]) {
					stk.push(t);
				}
			}
		}
		
		System.out.println();
		for(int i=0; i<=n; i++) {
			Collections.sort(graph.get(i));
		}
		
		// bfs
		que.offer(v);
		bfs_visited[v] = true;
		while(!que.isEmpty()) {
			int top = que.poll();
			System.out.print(top+" ");
			for(Integer t:graph.get(top)) {
				if(!bfs_visited[t]) {
					bfs_visited[t]=true;
					que.offer(t);
				}
			}
		}
	}
}